fix(drizzle): reimplement Drizzle adapter

realmikesolo opened the pull request:
##  Reasoning

The previous Drizzle adapter for NextAuth was implemented incorrectly. I have rewritten it from the ground up to fix the issues and align it with Drizzle's best practices. The adapter passed all tests. Documentation was updated according to new functionality.

## 🧢 Checklist

- [X] Documentation
- [X] Tests
- [ ] Ready to be merged

realmikesolo committed:
update drizzle adapter

realmikesolo committed:
update adapter for drizzle: finish

realmikesolo committed:
Merge remote-tracking branch 'upstream/main'

AlexBlokh commented on packages/adapter-drizzle/src/index.ts:
```diff
@@ -71,22 +91,23 @@ import type { Adapter } from "@auth/core/adapters"
  *   userId: text("userId")
  *     .notNull()
  *     .references(() => users.id, { onDelete: "cascade" }),
- *   type: text("type").$type<AdapterAccount["type"]>().notNull(),
```
removed, no other adapter has that type specified and it's redundant

AlexBlokh commented on packages/adapter-drizzle/src/index.ts:
```diff
@@ -71,22 +91,23 @@ import type { Adapter } from "@auth/core/adapters"
+ *  id: text("id").primaryKey().$defaultFn(() => randomUUID()),
  *  name: text("name"),
- *  email: text("email").notNull(),
+ *  email: text("email").unique(),
```
not every provider has a mandatory email, for example `github`

ThangHuuVu replied:
for GitHub, we still get the email if it's null https://github.com/nextauthjs/next-auth/blob/edc4fe4b4294fe6eed083f7acf641314722a4618/packages/core/src/providers/github.ts#L156-L163

ndom91 reviewed (commented):
Hey so in general I appreciate the effort! 

But changing the name of the tables is a bugger breaking change.. we'd have to bump the major version and ideally provide some sort of migration / codemod to automate as much as possible.

realmikesolo commented:
> Hey so in general I appreciate the effort!
> 
> But changing the name of the tables is a bugger breaking change.. we'd have to bump the major version and ideally provide some sort of migration / codemod to automate as much as possible.

Got it. I will revert to the previous table names

realmikesolo committed:
drizzle adapter: fix table names

realmikesolo committed:
drizzle-adapter: fix docs

AlexBlokh commented on packages/adapter-drizzle/src/lib/mysql.ts:
```diff
@@ -2,263 +2,260 @@ import { and, eq } from "drizzle-orm"
-
-      return user
+    async getUserByEmail(email: string) {
+      return await client
```
redundant await

AlexBlokh commented on packages/adapter-drizzle/src/lib/mysql.ts:
```diff
@@ -2,263 +2,260 @@ import { and, eq } from "drizzle-orm"
+
+      await client.insert(sessionsTable).values({ ...data, id })
 
       return await client
```
redundant await

AlexBlokh commented on packages/adapter-drizzle/src/lib/mysql.ts:
```diff
@@ -2,263 +2,260 @@ import { and, eq } from "drizzle-orm"
+      await client.update(usersTable).set(data).where(eq(usersTable.id, data.id))
 
-      return await client
+      const result =  await client
```
`const [result]` instead of `.then(...)`

AlexBlokh commented on packages/adapter-drizzle/package.json:
```diff
@@ -48,14 +49,13 @@
     "postgres": "^3.3.4",
     "tsx": "^4.7.0"
   }
-}
```
fix \n

realmikesolo committed:
drizzle adapter: edits

AlexBlokh commented:
@ndom91 we're ready for review and to be merged!

ThangHuuVu commented:
wow, thanks a lot for the contribution @realmikesolo @AlexBlokh , we'll look into it! cc @ndom91

balazsorban44 commented on packages/adapter-drizzle/package.json:
```diff
@@ -19,7 +19,8 @@
       "types": "./index.d.ts",
-      "import": "./index.js"
+      "import": "./index.js",
+      "require": "./index.js"
```
True, all Auth.js packages are ESM-only

realmikesolo replied:
The `require` line was added to ensure that the default tables imported from the `adapter-drizzle` package can be used with the `drizzle-kit` commands, such as `generate` and `push`. These commands are used for applying changes to the database. Without it, users will encounter a `no exports main defined` error when attempting to use `drizzle-kit` commands.

realmikesolo commented on packages/adapter-drizzle/src/index.ts:
````diff
@@ -48,8 +51,25 @@ import type { Adapter } from "@auth/core/adapters"
+ * If you want to use your own tables, you can pass them as a second argument
  * :::
+ * 
+ * ```ts title="pages/api/auth/[...nextauth].ts"
+ * import NextAuth from "next-auth"
+ * import GoogleProvider from "next-auth/providers/google"
+ * import { DrizzleAdapter } from "@auth/drizzle-adapter"
+ * import { db, accounts, sessions, users, verificationTokens } from "./schema"
+ *
+ * export default NextAuth({
+ *   adapter: DrizzleAdapter(db, { usersTable: users, accountsTable: accounts, sessionsTable: sessions, verificationTokensTable: verificationTokens }),
+ *   providers: [
+ *     GoogleProvider({
+ *       clientId: process.env.GOOGLE_CLIENT_ID,
+ *       clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+ *     }),
+ *   ],
+ * })
+ * ```
````
got it, I will update js docs

balazsorban44 committed:
chore(examples): move Docker files to Next.js template

We would like to have feature parity of Auth.js when hosted either on something like Vercel or in Docker. Moving the docker config to the same example app we host on Vercel and deploying it via Docker will make this easier to verify

balazsorban44 committed:
chore(examples): use `with-docker` dockerfile config

balazsorban44 committed:
chore(examples): simplify Dockerfile

noam-honig committed:
chore(express): update README.md example - .cts to .cjs (#10504)

Update README.md fixed .cts to .cjs

Co-authored-by: Nico Domino <yo@ndo.dev>

balazsorban44 committed:
Merge branch 'main' of github.com:nextauthjs/next-auth

balazsorban44 committed:
chore(examples): use npm in Dockerfile

balazsorban44 committed:
chore(examples): tweak

balazsorban44 committed:
chore(examples): drop `HOSTNAME`

balazsorban44 committed:
chore(examples): revert

realmikesolo committed:
Merge remote-tracking branch 'upstream/main'

realmikesolo committed:
drizzle-adapter: edits

ndom91 committed:
Merge branch 'main' into main

ndom91 commented:
@realmikesolo out of curiosity, what was wrong exactly about the previous implementation and what did you do differently?

AlexBlokh commented:
@ndom91 tables were hardcoded within library and anything you'd redeclare with `$` functions would be ignored on inserts/updates/etc

realmikesolo committed:
drizzle adapter: remove redundant connections in tests

ndom91 committed:
Merge branch 'main' into main

ndom91 commented:
So in general I think this is good to go - tests all passing for me locally. But there's a lot of formatting / syntax issues. I tried to run prettier over everythign and push it to your branch, but it looks like I'm not allowed.

Can you open up your fork's rights? Or format it yourself? Thanks! :pray:

realmikesolo committed:
drizzle adapter: fix syntax issues

ndom91 committed:
Merge branch 'main' into main

ndom91 committed:
Merge branch 'main' into main

themixednuts commented:
Thank you for this PR, I was literally just looking into this myself to do the same thing and wondered why there was hard-coded tables without any schema passthrough. Awesome work!

ndom91 commented:
@realmikesolo could you take a look at the `pnpm-lock` merge conflict :pray:

realmikesolo committed:
Merge remote-tracking branch 'upstream/main'

realmikesolo committed:
Merge remote-tracking branch 'upstream/main'

ndom91 committed:
Merge branch 'main' into main

AlexBlokh commented:
seems like we're good

ndom91 committed:
Merge branch 'main' into main

ndom91 commented:
Merging shortly :pray:

ndom91 committed:
Merge branch 'main' into main

juliusmarminge commented on packages/adapter-drizzle/src/lib/mysql.ts:
```diff
@@ -2,263 +2,282 @@ import { and, eq } from "drizzle-orm"
   return {
-    async createUser(data) {
-      const id = crypto.randomUUID()
+    async createUser(data: Omit<AdapterUser, "id">) {
+      const id = randomUUID()
```
could we change this so we can leverage drizzle's `.defaultFn()` here and let users have control of the id generation?

juliusmarminge replied:
you have for the sessions table so don't see why users table should be any different

juliusmarminge replied:
ref https://github.com/nextauthjs/next-auth/pull/10479/files#r1573284149

realmikesolo replied:
Hi, @juliusmarminge. As MySQL doesn't have `returning` clause, generated `id` was used to find the row after insertion. But you are right, we can find inserted row by `email` for user and by `sessionToken` for session and remove redundant `id` generation in the code.

ndom91 replied:
@realmikesolo will you remove these redundant ID's in your follow up PR as well?

realmikesolo replied:
yep @ndom91

juliusmarminge commented on packages/adapter-drizzle/src/index.ts:
```diff
@@ -223,22 +261,26 @@ import type { Adapter } from "@auth/core/adapters"
  * export const sessions = sqliteTable("session", {
- * sessionToken: text("sessionToken").notNull().primaryKey(),
+ * id: text("id").primaryKey().$defaultFn(() => randomUUID())
+ * sessionToken: text("sessionToken").notNull().unique(),
  * userId: text("userId")
  *   .notNull()
  *   .references(() => users.id, { onDelete: "cascade" }),
  * expires: integer("expires", { mode: "timestamp_ms" }).notNull(),
- * })
+ * }, (table) => ({
+ *  userIdIdx: index('Session_userId_index').on(table.userId)
+ * }))
```
this is also a breaking change for the underlying database - why was this made?

realmikesolo replied:
Agree with you. It was a mistake to add indexes for foreign keys. I will remove them in new PR

ndom91 replied:
@realmikesolo thanks for stepping up to remove those, feel free to tag me in the PR so we can get it merged quickly :pray:

juliusmarminge commented on packages/adapter-drizzle/src/lib/mysql.ts:
```diff
@@ -2,263 +2,282 @@ import { and, eq } from "drizzle-orm"
-    })
-  )
+export const mysqlUsersTable = mysqlTable("user" as string, {
+  id: varchar("id", { length: 255 })
+    .primaryKey()
+    .$defaultFn(() => randomUUID()),
```
so this just wont have any affect since the `createUser` function always generates an id with randomUUID()

lucasfeliciano commented:
Amazing! Great direction this is going! :) 

Would be possible to make the column names also customizable? 

![image](https://github.com/nextauthjs/next-auth/assets/968014/5c67ceab-6c43-4d58-b78f-5a0bc045e171)

Thinking the same thing about the field type.
For fields that are not coupled with the implementation might be nice to specify the type.

ndom91 commented:
@realmikesolo thanks for commenting and offering to fix a few of these issues, theres some discussion on this adapter / these issues in the discord as well if you'd like to join: 

- Invite: https://discord.authjs.dev
- Drizzle Adapter Thread: https://discord.com/channels/1200116961590399008/1231534836859146241

Also we merged the new docs while this PR was being worked on, the docs are now at `/docs/pages/getting-started/adapters/drizzle.mdx` :pray:

realmikesolo commented:
Thank you @ndom91!
